icommands.c

```diff
@@ -98,6 +103,173 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
+   * however for other menus try generic first. */
+  if ((menu != MENU_PAGER) && (menu != MENU_EDITOR) && (menu != MENU_GENERIC))
+  {
+    fn_name = mutt_get_func(OpGeneric, map->op);
+  }
+
+  /* if it's one of the menus above or generic doesn't find
+   * the function, try with it's own menu. */
+  if (!fn_name)
+  {
+    const struct Binding *bindings = km_get_table(menu);
+    if (!bindings)
+      return;
+
+    fn_name = mutt_get_func(bindings, map->op);
+  }
+
+  fprintf(fpout, "bind %s %s %s\n", menu_name, key_binding, fn_name);
+}
+
+/**
+ * bind_dump_menu - Dumps all the binds/macros of a menu into a file.
+ * @param fpout     Output file
+ * @param menu      Menu id to dump
+ * @param menu_name Name of the menu to dump
+ * @param bind      If true it's :bind icommand, else :macro
+ */
+static void dump_menu(FILE *fpout, const int menu, const char *menu_name, bool bind)
+{
+  struct Keymap *map = NULL, *next = NULL;
+
+  for (map = Keymaps[menu]; map; map = next)
+  {
+    next = map->next;
+
+    if (!bind && map->op == OP_MACRO)
+      dump_macro(fpout, menu_name, map);
+    else if (bind && map->op != OP_MACRO)
+      dump_bind(fpout, menu, menu_name, map);
+  }
+}
+
+/**
+ * icmd_bind - Parse 'bind' and 'macro' commands - Implements ::icommand_t
+ */
+static enum CommandResult icmd_bind(struct Buffer *buf, struct Buffer *s,
+                                    unsigned long data, struct Buffer *err)
+{
+  FILE *fpout = NULL;
+  bool valid_menu = false;
+  char tempfile[PATH_MAX];
+
+  if (!MoreArgs(s))
+    mutt_buffer_strcpy(buf, "all");
+  else
+    mutt_extract_token(buf, s, 0);
+
+  if (MoreArgs(s))
+  {
+    /* More arguments potentially means the user is using the
+     * ::command_t :bind command thus we delegate the task. */
+    return MUTT_CMD_ERROR;
+  }
+
+  mutt_mktemp(tempfile, sizeof(tempfile));
+  fpout = mutt_file_fopen(tempfile, "w");
+  if (!fpout)
+  {
+    mutt_buffer_printf(err, _("Could not create temporary file %s"), tempfile);
+    return MUTT_CMD_ERROR;
+  }
+
+  const bool bind = data == 0;
+  const bool dump_all = mutt_str_strcasecmp(buf->data, "all") == 0;
+  for (int i = 0; i < MENU_MAX; i++)
+  {
+    const char *menu_name = mutt_map_get_name(i, Menus);
+
+    if (dump_all || mutt_str_strcasecmp(buf->data, menu_name) == 0)
+    {
+      fflush(fpout);
+      const long file_size = mutt_file_get_size(tempfile);
+
+      valid_menu = true;
+      dump_menu(fpout, i, menu_name, bind);
+
+      /* If initial size is different than actual size, menu wasn't
+       * empty. If so we add a newline for readability between menus. */
+      fflush(fpout);
+      if (file_size != mutt_file_get_size(tempfile) && i < MENU_MAX - 1)
+        fputs("\n", fpout);
+
+      if (!dump_all)
+        break;
+    }
+  }
+
+  mutt_file_fclose(&fpout);
+
+  if (!valid_menu)
+  {
+    mutt_buffer_printf(err, _("%s: no such menu"), buf->data);
+    return MUTT_CMD_ERROR;
+  }
+  else if (mutt_file_check_empty(tempfile) == 1)
```

Comment:
Shouldn't this comparison be `== 0`?

`mutt_file_check_empty()` has three returns: 1 if the file has some in it, 0 if empty, or -1 if error. 

If there are no bindings/macros, the file would be empty, correct?

Reply:
Yep, I talked with @flatcap about this the other day on the IRC channel, turns out that `mutt_file_check_empty()` comment is incorrect.

If you take a look at the `mutt_file_check_empty()`'s return condition (`return st.st_size == 0;`) it should return 1 if empty

Reply:
>  it should return 1 if empty

Ah, I should know better than to assume documentation is 100% accurate.